adr/0677-css-encapsulation.md

```diff
@@ -0,0 +1,65 @@
+
+Over the past couple of months, we've experimented with bringing encapsulated CSS to Rails with ViewComponent, with the hope of addressing several key problems:
+
+### Scoping
+
+It's trivial to write CSS that has unintended consequences. By scoping styles using hashed class name selectors, this problem is unlikely.
+
+### Critical CSS
+
+In at least one case, we only use 3% of our main CSS bundle on a page, meaning a user downloads over 400kb of data they aren't using. By only serving the CSS for ViewComponents rendered in the current request, we can deliver only the CSS necessary to render a page.
+
+### Dead code elimination
+
+It can be hard to know if it's safe to remove CSS from a codebase. With encapsulation, it's safe to delete a component's CSS.
+
+## Decision
+
+We will implement significant portions of the long-standing [CSS Modules spec](https://github.com/css-modules/css-modules) natively in ViewComponent for [encapsulated styles](https://medium.com/seek-blog/the-end-of-global-css-90d2a4a06284):
+
+### No Webpacker
+
+The proposed design avoids Webpacker and other build tools. Instead, it uses a Ruby implementation of CSS Modules built by @rmosolgo, enabling greater portability and a convention-over-configuration experience that Rails developers expect.
+
+### No bundles
+
+At least for now, the proposed design eschews the concept of shipping CSS in bundles.
+
+### No global syntax
+
+In keeping with the existing ViewComponent encapsulation mental model, all styles are locally scoped by default, with no global escape hatch.
+
+### No SCSS
+
+We're not very sure about this one, but we're playing with the idea of just using plain CSS, at least to start.
+
+## Alternatives we considered
+
+### CSS-in-Ruby
+
+Much like our issue with [inlining ERB inside ViewComponent Ruby files](https://github.com/github/view_component/commit/5d3806a2a9ec0187574ac6c64a0ae90e655691ae#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R173), writing CSS inline (like Styled Components and others do) raised similar concerns over compatibility with Ruby 3 type systems.
+
+### Ruby-in-CSS
+
+We considered allowing for evaluation of Ruby inside CSS files (at compile, runtime, or both) but instead decided to lean on [CSS custom properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties) for injecting dynamic values.
+
+### Bundle splitting
+
+We attempted to split our CSS into multiple smaller bundles without significant success. We found it hard to be confident that such changes weren't breaking our UI in subtle ways. Ensuring heavily reused templates always rendered with their corresponding styles proved to be practically impossible.
+
+## Consequences
+
+ViewComponent authors will write encapsulated CSS, avoiding several common errors and resulting in more resilient user interfaces.
```

Comment:
I'm not sure where the right place is but I am left wondering how a project like this one that relies on Primer CSS will work. Will component authors have to pull in (duplicate) every class their component depends or is there some set of global primer css classes we get "for free"?

Reply:
@srt32 good question. I didn't mean for this statement to be so far-reaching. I've updated it to _ViewComponent authors will be able to write encapsulated CSS_